src/RateLimiting/src/ConcurrencyLimiter.cs

```diff
@@ -134,25 +193,43 @@ private void Release(int releaseCount)
 
         private class ConcurrencyLease : RateLimitLease
         {
-            private static readonly IEnumerable<string> Empty = new string[0];
-
             private bool _disposed;
             private readonly ConcurrencyLimiter? _limiter;
             private readonly int _count;
+            private readonly string? _reason;
 
-            public ConcurrencyLease(bool isAcquired, ConcurrencyLimiter? limiter, int count)
+            public ConcurrencyLease(bool isAcquired, ConcurrencyLimiter? limiter, int count, string? reason = null)
             {
                 IsAcquired = isAcquired;
                 _limiter = limiter;
                 _count = count;
+                _reason = reason;
+
+                // No need to set the limiter if count is 0, Dispose will noop
+                Debug.Assert(count == 0 ? limiter is null : true);
             }
 
             public override bool IsAcquired { get; }
 
-            public override IEnumerable<string> MetadataNames => Empty;
+            public override IEnumerable<string> MetadataNames => Enumerable();
+
+            private IEnumerable<string> Enumerable()
+            {
+                if (_reason is null)
+                {
+                    yield break;
+                }
+
+                yield return MetadataName.ReasonPhrase.Name;
```

Comment:
```suggestion
                if (_reason is not null)
                {
                    yield return MetadataName.ReasonPhrase.Name;
                }
```

This will allow for future additions without having to refactor this code.